packages/rum-react/src/get-with-transaction.js

```diff
@@ -76,17 +76,39 @@ function getWithTransaction(apm) {
        */
       if (
         !isReactClassComponent(Component) &&
-        typeof React.useEffect === 'function'
+        typeof React.useEffect === 'function' &&
+        typeof React.useState === 'function'
       ) {
         ApmComponent = function ApmComponent(props) {
-          const transaction = apm.startTransaction(name, type, {
-            canReuse: true
-          })
+          /**
+           * We start the transaction as soon as the ApmComponent gets rendered
+           * so that we can capture all the effects inside child components
+           *
+           * The reason why we have this transaction inside setState is that we don't
+           * want this piece of code to run on every render instead we want to
+           * start the transaction only on component mounting
+           */
+          const [transaction] = React.useState(() =>
+            apm.startTransaction(name, type, {
+              canReuse: true
+            })
+          )
 
+          /**
+           * React guarentess the parent component effects are run after the child components effects
+           * So once all the child components effects are run, we run the detectFinish logic
+           * which ensures if the transaction can be completed or not.
+           */
           React.useEffect(() => {
             transaction && transaction.detectFinish()
-            return () => transaction && transaction.detectFinish()
-          })
+            return () => {
+              /**
+               * Incase the transaction is never ended, we close the
+               * transaction when component unmounts
+               */
+              transaction && transaction.detectFinish()
```

Comment:
should this be `transaction.end()`? since even if the transaction has scheduled tasks while the component unmount, it doesn't make sense to keep it open. For the class components we are also using `detectFinish` so, not strictly related to this PR.

Reply:
Yeah you are  right, will change it to `.end` in both places.